### PR TITLE
fix WalkTime for negativ positions

### PR DESCRIPTION
**Pull request by C0D3D3V**
I am not sure if there is a more elegant solution, the problem is that fmod in cpp does not return a positive value. maybe we can just use a mod function that always returns positive values.

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [x] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by Jupeyy** [comment]
Just adding an offset of 200 tiles breaks smth?

BTW is this bug visible in game?

**Comment by C0D3D3V** [comment]
It is the moon walk bug. If you walk left of the border the walk animation is reversed. 
The calculation is not based on tiles but on units. So every 3.125 tiles the animation does one complete round. 
I made this WIP because I had no more time yesterday and wanted to delete the absolute() an second fmod() because both function do nothing now.

**Comment by Jupeyy** [comment]
OK, world border sounds fixable by an offset tho, but haven't looked deeper into it. Was just curious;)

**Comment by heinrich5991** [comment]
```c
#include <math.h>
#include <stdio.h>

float fmodf2(float x, float y) {
	float result = fmodf(x, y);
	if (result < 0) {
		result += y;
	}
	return result;
}

int main() {
	printf("29%%32=%f\n", fmodf(29.0f, 32.0f));
	printf("-3%%32=%f\n", fmodf(-3.0f, 32.0f));
	printf("\n");
	printf("29%%32=%f\n", fmodf2(29.0f, 32.0f));
	printf("-3%%32=%f\n", fmodf2(-3.0f, 32.0f));
	return 0;
}
```

```
29%32=29.000000
-3%32=-3.000000

29%32=29.000000
-3%32=29.000000
```

Use that function, I'd say.

File: src/game/client/components/players.cpp [hunk 1]
```diff
@@ -397,7 +397,8 @@ void CPlayers::RenderPlayer(
 	bool WantOtherDir = (Player.m_Direction == -1 && Vel.x > 0) || (Player.m_Direction == 1 && Vel.x < 0);
 
 	// evaluate animation
-	float WalkTime = fmod(absolute(Position.x), 100.0f) / 100.0f;
+	float WalkTime = Position.x >= 0 ? fmod(absolute(Position.x), 100.0f) / 100.0f : fmod(fmod(Position.x, 100.0f) + 100.0f, 100.0f) / 100.0f;
+	;
```
Comment (heinrich5991) [inline src/game/client/components/players.cpp:401]:
```suggestion
	float WalkTime = Position.x >= 0 ? fmod(Position.x, 100.0f) / 100.0f : (fmod(Position.x, 100.0f) + 100.0f) / 100.0f;
```

**Comment by C0D3D3V** [comment]
@heinrich5991 yes I also found this post yesterday https://stackoverflow.com/a/58118871
that also suggests such a function.

**Comment by C0D3D3V** [comment]
Can be tested on the map Forgotten Lands on the spawn

**Comment by def-** [comment]
bors r+
Works

**Pull request merged**